hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/EventloopBuilder.java

```diff
@@ -0,0 +1,174 @@
+
+import com.hazelcast.internal.util.ThreadAffinity;
+
+import java.util.concurrent.ThreadFactory;
+import java.util.function.Supplier;
+
+import static com.hazelcast.internal.tpc.util.Preconditions.checkNotNegative;
+import static com.hazelcast.internal.tpc.util.Preconditions.checkNotNull;
+import static com.hazelcast.internal.tpc.util.Preconditions.checkPositive;
+import static java.lang.System.getProperty;
+
+/**
+ * A factory for {@link Eventloop} instances.
+ */
+public abstract class EventloopBuilder {
+    public static final String NAME_LOCAL_TASK_QUEUE_CAPACITY = "hazelcast.tpc.localTaskQueue.capacity";
+    public static final String NAME_CONCURRENT_TASK_QUEUE_CAPACITY = "hazelcast.tpc.concurrentTaskQueue.capacity";
+    public static final String NAME_SCHEDULED_TASK_QUEUE_CAPACITY = "hazelcast.tpc.scheduledTaskQueue.capacity";
+    public static final String NAME_BATCH_SIZE = "hazelcast.tpc.batch.size";
+    public static final String NAME_CLOCK_REFRESH_PERIOD = "hazelcast.tpc.clock.refreshPeriod";
+    public static final String NAME_EVENTLOOP_SPIN = "hazelcast.tpc.eventloop.spin";
+
+    private static final int DEFAULT_LOCAL_QUEUE_CAPACITY = 65536;
+    private static final int DEFAULT_CONCURRENT_QUEUE_CAPACITY = 65536;
+    private static final int DEFAULT_SCHEDULED_TASK_QUEUE_CAPACITY = 4096;
+    private static final int DEFAULT_BATCH_SIZE = 64;
+    private static final int DEFAULT_CLOCK_REFRESH_INTERVAL = 16;
+    private static final boolean DEFAULT_SPIN = false;
+
+    protected final EventloopType type;
+    Supplier<Scheduler> schedulerSupplier = NopScheduler::new;
+    Supplier<String> threadNameSupplier;
+    ThreadAffinity threadAffinity = ThreadAffinity.newSystemThreadAffinity("hazelcast.tpc.eventloop.affinity");
+
+    ThreadFactory threadFactory = Thread::new;
+    boolean spin;
+    int localTaskQueueCapacity;
+    int concurrentTaskQueueCapacity;
+    int scheduledTaskQueueCapacity;
+    int batchSize;
+    int clockRefreshPeriod;
+
+    protected EventloopBuilder(EventloopType type) {
+        this.type = checkNotNull(type);
+        this.localTaskQueueCapacity = Integer.getInteger(NAME_LOCAL_TASK_QUEUE_CAPACITY, DEFAULT_LOCAL_QUEUE_CAPACITY);
+        this.concurrentTaskQueueCapacity = Integer.getInteger(NAME_CONCURRENT_TASK_QUEUE_CAPACITY, DEFAULT_CONCURRENT_QUEUE_CAPACITY);
+        this.scheduledTaskQueueCapacity = Integer.getInteger(NAME_SCHEDULED_TASK_QUEUE_CAPACITY, DEFAULT_SCHEDULED_TASK_QUEUE_CAPACITY);
+        this.batchSize = Integer.getInteger(NAME_BATCH_SIZE, DEFAULT_BATCH_SIZE);
+        this.clockRefreshPeriod = Integer.getInteger(NAME_CLOCK_REFRESH_PERIOD, DEFAULT_CLOCK_REFRESH_INTERVAL);
+        this.spin = Boolean.parseBoolean(getProperty(NAME_EVENTLOOP_SPIN, "" + DEFAULT_SPIN));
+    }
+
+    /**
+     * Creates an Eventloop based on the configuration of this {@link EventloopBuilder}.
+     *
+     * @return the created Eventloop.
+     */
+    public abstract Eventloop create();
```

Comment:
Minor: Maybe rename to build after renaming the class?

Reply:
Done. newBuilder.